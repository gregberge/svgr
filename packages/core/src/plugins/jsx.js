import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
