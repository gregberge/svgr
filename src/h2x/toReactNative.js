const elementToComponent = {
  svg: 'Svg',
  circle: 'Circle',
  clipPath: 'ClipPath',
  ellipse: 'Ellipse',
  g: 'G',
  linearGradient: 'LinearGradient',
  radialGradient: 'RadialGradient',
  line: 'Line',
  path: 'Path',
  pattern: 'Pattern',
  polygon: 'Polygon',
  polyline: 'Polyline',
  rect: 'Rect',
  symbol: 'Symbol',
  text: 'Text',
  textPath: 'TextPath',
  tspan: 'TSpan',
  use: 'Use',
  defs: 'Defs',
  stop: 'Stop',
}

const componentToElement = Object.keys(elementToComponent).reduce(
  (map, key) => ({ ...map, [elementToComponent[key]]: key }),
  {},
)

const toReactNative = () => ({
  visitor: {
    JSXElement: {
      enter(path, state) {
        // Replace element by react-native-svg components
        const component = elementToComponent[path.node.name]
        if (component) {
          path.node.name = component
          state.reactNativeSvgReplacedComponents =
            state.reactNativeSvgReplacedComponents || new Set()
          state.reactNativeSvgReplacedComponents.add(component)
          return
        }

        // Remove element if not supported
        if (!componentToElement[path.node.name]) {
          state.unsupportedComponents = state.unsupportedComponents || new Set()
          state.unsupportedComponents.add(path.node.name)
          path.remove()
        }
      },
    },
  },
})

export default toReactNative
