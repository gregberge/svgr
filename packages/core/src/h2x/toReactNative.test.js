import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import toReactNative from './toReactNative'

describe('toReactNative', () => {
  it('should take svg and convert it to react-native', () => {
    const result = transform(
      `
      <svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, toReactNative()] },
    )

    expect(result).toMatchSnapshot()
  })
})
