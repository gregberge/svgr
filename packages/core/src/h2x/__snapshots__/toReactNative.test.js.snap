// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toReactNative should take svg and convert it to react-native 1`] = `
"<Svg>
  <G stroke=\\"#063855\\" strokeWidth={2}>
    <Path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </G>
</Svg>
"
`;
